### PR TITLE
Update SMARTS reset with start time and fix dt dependency.

### PR DESCRIPTION
**Pull request by Gamenot**
This provides the ability to move the simulator ahead on reset. And fixes an issue where the Trap manager was using time deltas instead of the current simulation time.

See #1407 for context.

### CODE REVIEW

**Review by qianyi-sun — APPROVED** [review]

File: smarts/core/smarts.py
Comment (saulfield) [inline smarts/core/smarts.py:365]: Perhaps it's worth adding a type annotation for the public interface? I see there are a few other places in this diff that are missing them as well, but I'll leave that up to you.
```suggestion
    def reset(self, scenario: Scenario, start_time: float = 0.0) -> Dict[str, Observation]:
```

**Review by saulfield — APPROVED** [review]

**Review by RutvikGupta — APPROVED** [review]

**Comment by YuechengLiu** [comment]
Hello @Gamenot , Thanks for your work!

How should I use this new inferface?  
Do I need to set the start_time of `ego_missions` to 0 with the new inferface?

**Comment by YuechengLiu** [comment]
Maybe SMARTS should provide a inferface to discover the mission of a vehicle for any given start_time (not only for the first-seen-time of the vehicle).

**Comment by Gamenot** [comment]
> Hello @Gamenot , Thanks for your work!
> 
> How should I use this new inferface? Do I need to set the start_time of `ego_missions` to 0 with the new inferface?
> 
> Does the interface change have an impact on existing examples? Such as https://github.com/huawei-noah/SMARTS/blob/master/examples/history_vehicles_replacement_for_imitation_learning.py

@YuechengLiu This interface in its current form should have no effect on the current history replacement example. It would only have an effect if you change the `start_time` to something other than 0. The `ego_missions` start time should be set to at or after the `start_time` provided to reset with this changeset or they may be skipped (and a warning given.)

The exact interface implications are the following:

```python
assert all(start_time <= (m.start_time + m.entry_tactic.wait_to_hijack_limit_s) for m in missions)
# Warning occurs with skipped missions that would run out of patience before `start_time` 
smarts.reset(scenario, start_time = start_time) 
```

**Comment by YuechengLiu** [comment]
@Gamenot 

https://github.com/huawei-noah/SMARTS/blob/3ef5650b04ac6fb7145cf4e23d5534d73e0929fc/examples/history_vehicles_replacement_for_imitation_learning.py#L206

According to my understanding, for example, we should remove the line here. Is this right?

**Comment by Gamenot** [comment]
> Maybe SMARTS should provide an interface to discover the mission of a vehicle for any given start_time (not only for the first-seen-time of the vehicle).

A new request seems to be a different issue. I assume the intention is something like the following:

```python
active_vehicle_ids = traffic_history.vehicle_ids_active_between(start_time, end_time)
used_vehicle_ids = custom_filter(active_vehicle_ids)
missions = {
    vehicle_id: mission_between_time(
      traffic_history,
      vehicle_id,
      start_time, # could be different value
      end_time # could be different value
    ) for vehicle_id in used_vehicle_ids
}
assert all((
        m.start_time == start_time 
        && isinstance(m.goal, PositionalGoal)
    ) for m in vehicle_missions.values()
)
```

**Comment by Gamenot** [comment]
I will do some additional tests tomorrow morning before merging this PR.

**Comment by YuechengLiu** [comment]
> > Maybe SMARTS should provide an interface to discover the mission of a vehicle for any given start_time (not only for the first-seen-time of the vehicle).
> 
> A new request seems to be a different issue. I assume the intention is something like the following:
> 
> ```python
> active_vehicle_ids = traffic_history.vehicle_ids_active_between(start_time, end_time)
> used_vehicle_ids = custom_filter(active_vehicle_ids)
> missions = {
>     vehicle_id: mission_between_time(
>       traffic_history,
>       vehicle_id,
>       start_time, # could be different value
>       end_time # could be different value
>     ) for vehicle_id in used_vehicle_ids
> }
> assert all((
>         m.start_time == start_time 
>         && isinstance(m.goal, PositionalGoal)
>     ) for m in vehicle_missions.values()
> )
> ```

This is exactly what I need.

File: smarts/core/smarts.py
Comment (sah-huawei) [inline smarts/core/smarts.py:396]: type hint

**Review by sah-huawei — COMMENTED** [review]
Should we use this in the `reset()` call in `history_vehicles_replacement_for_imitation_learning.py`?  (Currently, it uses the `history_start_time` variable to shift trajectories in time, but it might be cleaner to just start the simulation at the appropriate time.)

**Comment by Gamenot** [comment]
> > > Maybe SMARTS should provide an interface to discover the mission of a vehicle for any given start_time (not only for the first-seen-time of the vehicle).
> > ...
> 
> This is exactly what I need.

I have taken a look into it and I see why it is needed during testing. The exact interface I am going for right now is the following:

```python3
traffic_history = scenario.traffic_history
active_vehicles: Dict[id_str, HistoryVehicleSlice] = traffic_history.vehicle_slice_in_range(start_time, end_time)
# e.g. filter: v.end_time - v.start_time > 6 and v.start_time == start_time
used_vehicle_ids = custom_filter(active_vehicles) 
missions = {
    vehicle_id: scenario.history_mission_for_time(
      vehicle_id,
      active_vehicles[vehicle_id].start_time, # could be different value
      active_vehicles[vehicle_id].end_time, # value > start_time
    ) for vehicle_id in used_vehicle_ids
}
```

**Pull request merged**
